packages/react-ai/src/hooks/useAIGeneration.tsx

```diff
@@ -45,31 +55,37 @@ export function createUseAIGeneration<
     Key extends keyof AIGenerationClient<Schema>['generations'],
   >(
     routeName: Key
-  ) => {
+  ): [
+    state: GenerateState<Schema[Key]['returnType']>,
+    handleAction: (input: Schema[Key]['args']) => void,
+  ] => {
+    const [graphqlErrors, setGraphqlErrors] = React.useState<
+      GraphQLFormattedError[] | undefined
+    >(undefined);
+
     const handleGenerate = (
       client.generations as AIGenerationClient<Schema>['generations']
     )[routeName];
 
     const updateAIGenerationStateAction = async (
       _prev: Schema[Key]['returnType'],
       input: Schema[Key]['args']
-    ): Promise<
-      Schema[Key]['returnType'] & { graphqlErrors?: GraphQLFormattedError[] }
-    > => {
+    ): Promise<Schema[Key]['returnType']> => {
       const result = await handleGenerate(input);
 
-      // handleGenerate returns a Promised wrapper around Schema[Key]['returnType'] which includes data, errors, and clientExtensions
-      // The type of data is Schema[Key]['returnType'] which useDataState also wraps in a data return
-      // TODO: follow up with how to type handleGenerate to properly return the promise wrapper shape
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-      const data = (result as any).data as Schema[Key]['returnType'];
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
-      const graphqlErrors = (result as any).errors;
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-assignment
-      return { ...data, ...(graphqlErrors ? { graphqlErrors } : {}) };
+      const { data, errors } = result as SingularReturnValue<
+        Schema[Key]['returnType']
+      >;
+      setGraphqlErrors(errors);
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+      return data;
     };
 
-    return useDataState(updateAIGenerationStateAction, {});
+    const [data, handler] = useDataState(
+      updateAIGenerationStateAction,
+      undefined
+    );
+    return [{ ...data, graphqlErrors }, handler];
```

Comment:
Can simplify a bit and remove the `graphqlErrors` state by returning `result` directly from `updateAIGenerationStateAction` and restructuring the hook return output, something like:
```ts
    const { data, errors } =
      (result?.data as SingularReturnValue<Schema[Key]['returnType']>) ?? {};

    const output: [
      state: GenerateState<Schema[Key]['returnType']>,
      handleAction: (input: Schema[Key]['args']) => void,
    ] = [{ ...result, data, graphqlErrors: errors }, handler];
```

Reply:
Ah damn this is much cleaner, let me try that and ping you :)